core-processor/src/processor.rs

```diff
@@ -233,87 +231,192 @@ pub fn precharge(
     }
 }
 
-/// Prepares environment for the execution of a program.
-/// Checks if there is a required export and tries to charge for code and memory pages.
-/// Returns either `PreparedMessageExecutionContext` for `process` or an array of journal notes.
-/// See `PrepareResult` for details.
-pub fn prepare(
+/// Charge a message for fetching the length of the binary code.
+pub fn precharge_for_code_length(
     block_config: &BlockConfig,
-    execution_context: MessageExecutionContext,
-) -> PrepareResult {
+    dispatch: PrechargedDispatch,
+    destination_id: ProgramId,
+    executable_data: Option<ExecutableActorData>,
+) -> PrechargeResult<ContextChargedForCodeLength> {
     use executor::ChargeForBytesResult;
 
-    let read_per_byte_cost = block_config.read_per_byte_cost;
     let read_cost = block_config.read_cost;
-    let max_reservations = block_config.max_reservations;
-
-    let MessageExecutionContext {
-        actor,
-        precharged_dispatch,
-        origin,
-        subsequent_execution,
-    } = execution_context;
-    let Actor {
-        balance,
-        destination_program: program_id,
-        executable_data,
-    } = actor;
 
-    let (dispatch, mut gas_counter, mut gas_allowance_counter) = precharged_dispatch.into_parts();
+    let (dispatch, mut gas_counter, mut gas_allowance_counter) = dispatch.into_parts();
 
     let actor_data = match check_is_executable(executable_data, &dispatch) {
         Err(status_code) => {
-            return PrepareResult::Error(process_non_executable(dispatch, program_id, status_code));
+            return Err(process_non_executable(
+                dispatch,
+                destination_id,
+                status_code,
+            ));
         }
         Ok(data) => data,
     };
 
     if !actor_data.code_exports.contains(&dispatch.kind()) {
-        return PrepareResult::WontExecute(process_success(
+        return Err(process_success(
             SuccessfulDispatchResultKind::Success,
-            DispatchResult::success(dispatch, program_id, gas_counter.into()),
+            DispatchResult::success(dispatch, destination_id, gas_counter.into()),
         ));
     }
 
+    match executor::charge_gas_for_bytes(read_cost, &mut gas_counter, &mut gas_allowance_counter) {
```

Comment:
Use here `charge_gas_for_bytes` is also misleading. Better to rename this function to "charge_gas"

Reply:
Agree but there is already the function with that name so I would leave things as is.

Reply:
What if we go with sth like `charge_per_byte`?